### ISSUE TITLE
Support the build cache and parallel test execution with the JaCoCo plugin applied

### ISSUE THREAD

**Issue by JLLeitschuh** [gradle/gradle#5269]
### Update
The wider implications of this bug is captured well here:
https://github.com/gradle/gradle/issues/5269#issuecomment-390489401

### Expected Behavior
If you use the Jacoco plugin with it's default configuration, it should not break the `Test` task's ability to be cached.

### Current Behavior
Applying the `Jacoco` plugin to your project makes the `Test` task not support caching.

https://github.com/gradle/gradle/blob/76387d95bd2d785eb76991d252102b9546551696/subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java#L131-L136

https://docs.gradle.org/current/dsl/org.gradle.testing.jacoco.plugins.JacocoTaskExtension.html#org.gradle.testing.jacoco.plugins.JacocoTaskExtension:append

> Whether or not data should be appended if the destinationFile already exists. Defaults to true.

### Context
It makes sense that if you have custom configuration for the `Jacoco` plugin you might prevent Build Caching from working, but in it's default configuration it should work.

### Steps to Reproduce (for bugs)
Add the `Jacoco` plugin to a project and be using the Build Cache.

### Workaround
Using the Kotlin DSL:
```kotlin
tasks.withType<Test> {
    extensions.configure(typeOf<JacocoTaskExtension>()) {
        isAppend = false
    }
}
```

### Your Environment
```
------------------------------------------------------------
Gradle 4.7
------------------------------------------------------------

Build time:   2018-04-18 09:09:12 UTC
Revision:     b9a962bf70638332300e7f810689cb2febbd4a6c

Groovy:       2.4.12
Ant:          Apache Ant(TM) version 1.9.9 compiled on February 2 2017
JVM:          1.8.0_92 (Oracle Corporation 25.92-b14)
OS:           Mac OS X 10.13.3 x86_64

```

**Comment by wolfs** [comment]
We would love to change the default for `append` in the Jacoco plugin, but that would be a breaking change :(
So we need to wait for Gradle 5.0 to change the default to `append = false`, so caching would be enabled by default for the `Test` task with JaCoCo enabled.

**Comment by wolfs** [comment]
If you are willing to work on a solution in the JaCoCo plugin to support parallel test execution with caching and the Jacoco plugin enabled, then we would be happy to assist you in building the corresponding PR.

I think the basic approach would be to interpret `append = false` in the JaCoCo extension to mean:
- clean up the coverage file when the `Test` task starts
- pass `append = true` to the JaCoCo agent

We would need some test coverage making sure that parallel test execution works well with the Jacoco plugin and `append = true`.

When this works, we would set `append = false` in Gradle 5.0 and thereby enable caching of `Test` tasks with JaCoCo coverage by default.

**Comment by big-guy** [comment]
@wolfs If we're making a breaking change in 5.0, we may also just want to remove configuration options that don't make sense or make it difficult to do the right thing. 

Even though there's a `JacocoMerge` task that can take multiple JaCoCo reports and generate a single execution file, some builds accomplish the same thing by using the same output file.   This could be from multiple `Test` tasks or multiple executions of the same `Test` task with different parameters.

**Comment by JLLeitschuh** [comment]
@wolfs What about simply having each worker generate their own JaCoCo report?
As the Jacoco task is today, it already supports merging multiple JaCoCo execution report outputs into one.

We do that here at Plexxi to create our uber coverage report from all of our sub-projects.

```kotlin
val jacocoRootReport = task<JacocoReport>("jacocoRootReport") {
    group = LifecycleBasePlugin.VERIFICATION_GROUP
    description = "Generates code coverage report for all sub-projects."

    val jacocoReportTasks =
        subprojects
            .filter {
                // Filter out source sets that don't have tests in them
                // Otherwise, Jacoco tries to generate coverage data for tests that don't exist
                !it.java.sourceSets["test"].allSource.isEmpty
            }
            .map { it.tasks[jacocoTestResultTaskName] as JacocoReport }
    dependsOn(jacocoReportTasks)

    val allExecutionData = jacocoReportTasks.map { it.executionData }
    executionData(*allExecutionData.toTypedArray())

    // Pre-initialize these to empty collections to prevent NPE on += call below.
    additionalSourceDirs = files()
    sourceDirectories = files()
    classDirectories = files()

    junit5Projects.forEach { testedProject ->
        val sourceSets = testedProject.java.sourceSets
        this@task.additionalSourceDirs = this@task.additionalSourceDirs?.plus(files(sourceSets["main"].allSource.srcDirs))
        this@task.sourceDirectories += files(sourceSets["main"].allSource.srcDirs)
        this@task.classDirectories += files(sourceSets["main"].output)
    }

    reports {
        html.isEnabled = true
        xml.isEnabled = true
        csv.isEnabled = false
    }
}
```

**Comment by JLLeitschuh** [comment]
Is there a `release/5.0` branch where a PR of this nature could be pointed?

**Comment by JLLeitschuh** [comment]
@big-guy @w4tson The implications of this bug are a bit wider reaching than originally captured by this bug.

Gathering from what you've stated here, it seems that if you choose to have multiple workers for your tests, and you want to capture Jacoco code-coverage data, there is no way to also get build cache support.

So your options are, build cache support, or tests on multiple workers, but not both.

Does this merit another issue being opened as this implication seems to be of slightly wider concern than this issue originally tried to capture?

**Comment by wolfs** [comment]
@JLLeitschuh You are exactly right.

When using the java and jacoco plugins, users have the complicated and unclear choice of...

- Accepting the defaults and not being able to cache their Test execution at all.
- Reconfiguring `jacoco.append=false` so that `Test` will be cacheable. But this comes at the cost of not using parallel test execution.

It would be better if someone could have jacoco analysis, parallel test execution, Test cacheability -- all by default or easily configurable.

I wouldn't open a new issue but change the name of the current issue. I gave the rename a shot. WDYT?

**Comment by JLLeitschuh** [comment]
@wolfs Unfortunately, I don't think I'm going to have the time to fix this myself. Is there any possibility that someone from the team will get the chance to tackle this before `5.0`?

**Comment by wolfs** [comment]
According to [this issue](https://github.com/jacoco/jacoco/blob/b9d1b54e300318b470d9fedccc69d75187016444/org.jacoco.agent.rt/src/org/jacoco/agent/rt/internal/output/FileOutput.java#L69) `append = true` should work.

The build cache is disabled if you didn't explicitly enable it and it will be disabled for all `Test` tasks with the JaCoCo agent configured with `append=true`.

I don't know about any issue regarding using parallel forks and `append = true`.

Writing different report files per test worker is not supported, yet.

The build cache shouldn't have any effect on whether tests are executed or not. Are you talking about single tests within the same `Test` task or whole `Test` tasks not being executed?

**Comment by cal101** [comment]
Thanks for yourvanswer.

It's a test job of about 8000 tests. Last time less then 2000 tests were executed. So whole testsuites in junit jargon are missing.

**Comment by wolfs** [comment]
@cal101 That sounds like a bug. Could you open another issue with some instructions how to reproduce the problem?

**Comment by cal101** [comment]
@wolfs I will try to make a reproducible case but that's difficult. It's on a closed source legacy project. It may be easier for me trying to debug the issue. I will start trying to debug.

**Comment by wolfs** [comment]
Fixed via https://github.com/gradle/gradle/pull/6419.

**Comment by JLLeitschuh** [comment]
@wolfs Thank you for fixing this!!! Is `append` now simply ignored?

**Comment by rawasthi11** [comment]
Hi Every one , 
I have a scenario, where i want to execute multiple test cases on same code in parallel , issue in this if i have execute multiples test cases at a same , generated exec file don't have proper mapping of test case with java class.(as in background other java files also get loaded which are evoked by other test cases).

i want for one test case , only single exec file is generated having mapping of java classes which are touched by this tests case only. (this is possible , if i execute test cases in serial). In order to save time, i want to execute multiple test cases at same time, and in exec only those java classes covered which is touched by test case specific to it.

Is there any way to achieve this ?
